bundles/org.openhab.binding.mideaac/src/main/java/org/openhab/binding/mideaac/internal/cloud/Cloud.java

```diff
@@ -0,0 +1,334 @@
+
+        // .version(HttpVersion.HTTP_1_1)
+        request.agent("Dalvik/2.1.0 (Linux; U; Android 7.0; SM-G935F Build/NRD90M)");
+
+        if (!cloudProvider.proxied().isBlank()) {
+            request.header("Content-Type", "application/json");
+        } else {
+            request.header("Content-Type", "application/x-www-form-urlencoded");
+        }
+
+        request.header("secretVersion", "1");
+
+        // Add the sign to the header, different for proxied
+        if (!cloudProvider.proxied().isBlank()) {
+            String sign = security.newSign(json, random);
+            request.header("sign", sign);
+        } else {
+            if (!Objects.isNull(sessionId) && !sessionId.isBlank()) {
+                data.addProperty("sessionId", sessionId);
+            }
+            String sign = security.sign(url, data);
+            data.addProperty("sign", sign);
+            request.header("sign", sign);
+        }
+
+        request.header("random", random);
+
+        // If available, blank if not
+        request.header("accessToken", accessToken);
+
+        logger.debug("Request headers: {}", request.getHeaders().toString());
+
+        // Different formats for proxied
+        if (!cloudProvider.proxied().isBlank()) {
+            request.content(new StringContentProvider(json));
+        } else {
+            String body = Utils.getQueryString(data, false);
+            logger.debug("Request body: {}", body);
+            request.content(new StringContentProvider(body));
+        }
+
+        // POST the payload
+        @Nullable
+        ContentResponse cr = null;
+        try {
+            cr = request.send();
+        } catch (InterruptedException e) {
+            logger.warn("an interupted error has occurred{}", e.getMessage());
+        } catch (TimeoutException e) {
+            logger.warn("a timeout error has occurred{}", e.getMessage());
+        } catch (ExecutionException e) {
+            logger.warn("an execution error has occurred{}", e.getMessage());
+        }
+
+        if (cr != null) {
+            logger.debug("Response json: {}", cr.getContentAsString());
+            JsonObject result = Objects.requireNonNull(new Gson().fromJson(cr.getContentAsString(), JsonObject.class));
+
+            int code = -1;
+
+            if (result.get("errorCode") != null) {
+                code = result.get("errorCode").getAsInt();
+            } else if (result.get("code") != null) {
+                code = result.get("code").getAsInt();
+            } else {
+                errMsg = "No code in cloud response";
+                logger.warn("Error logging to Cloud: {}", errMsg);
+                return null;
+            }
+
+            String msg = result.get("msg").getAsString();
+            if (code != 0) {
+                errMsg = msg;
+                logger.warn("Error {} logging to Cloud: {}", code, msg);
```

Comment:
This logger shows that the interaction with this class and its callers can be improved.
How would the caller know that the login failed and route that to the thing lifecycle? 
 
Usually these type of classes throw a custom exception like `LoginFailedException` or `TimeOutException`. 
Either way it is good to make a distinction between transient and permanent problems so the caller can act on it.